feat: determine merges from base file imports

nawilson opened the pull request:
This feature enables use of a "base" file via the parameter "--base". It will recursively determine which Prisma schema files to merge based on the import statements in each file. The file specified as "base" will be the starting point.

Note: --base is mutually exclusive with --schemas. Parameter checking (via arg and package.json) updated to ensure either base or schemas is used, but not both at the same time.

ajmnz reviewed (changes requested):
Looks interesting! I still think this puts some unnecessary responsibility to the user because they might not be fully aware of which schemas are being picked up, and it forces them to maintain a kind of `index.prisma`. For example:

```prisma
// schema-1.prisma

import { Bar } from "./schema-2"

model Foo {
  id  Int   @id
  bar Bar[]
}

// schema-2.prisma

import { Foo } from "./schema-1"

model Bar {
  id  Int   @id
  foo Foo[]
}

// schema-3.prisma

import { Qux } from "./schema-4"

model Baz {
  id  Int   @id
  qux Qux[]
}

// schema-4.prisma

import { Baz } from "./schema-3"

model Qux {
  id  Int   @id
  foo Baz[]
}
```

What `--base` would you specify here? If you set it to `schema-1`, then 3 and 4 wouldn't get picked up. You'd probably need a file that has just import statements for the sake of "auto-resolving" to work.

Either way, I'm sure this will be useful for many users if they are aware of the perks and consequences.

Thanks!

nawilson commented:
> Looks interesting! I still think this puts some unnecessary responsibility to the user because they might not be fully aware of which schemas are being picked up, and it forces them to maintain a kind of `index.prisma`. 

That is exactly the purpose. Here are the workflow scenarios:

1. Single app repo:  In this workflow, the developer will have prisma files that all need to be deployed, as they're all built for the single app. In this case using the  `--schemas` globs to provide the prisma files is the best option.

2. Monorepo - multi app: In this workflow, the developer will have prisma files that can be shared across the multiple (and possibly different domain) apps that they have in their repo. For example in Nx, the developer may follow the directory structure (just an example):

```
├── apps
│   ├── app1
│   │   ├── app.prisma
│   ├── app2
│   │   ├── app.prisma
├── libs
│   ├── app1
│   │   ├── **/*/*.prisma
│   ├── app2
│   │   ├── **/*/*.prisma
│   ├── shared
│   │   ├── **/*/*.prisma
├── dist
```
The goal is to build 1 schema.prisma for each app. Each app may require a different set of tables from the `shared` directory.

Approach 1: `--schemas libs/shared/**/*.prisma`. All tables will be merged; not just the ones required for the app. 

Approach 2: Use multiple `--schemas` that are more specific. However, that would be time consuming and error prone for the developer to manually identify - especially if there are subsequent changes to the shared prisma files that introduce new dependencies.

Approach 3: Use `--base`. This allows the developer to identify, either directly or indirectly, all the required prisma files via imports. So by running `--base apps/app1/app.prisma` they will get only the tables required for that specific application.

ajmnz commented:
> That is exactly the purpose.

I understand the purpose and its value. In fact, I had this same scenario when building the VSCode extension, where there was no reasonable way of identifying different workspaces or apps. (it still is a limitation)

What I really meant was that, even when in a monorepo, the --base argument might not pick up all schemas inside that workspace because there's the possibility that not all will be linked between them, thus causing the resolver to "miss" some schemas that the user assumed would be picked up. The implications of this are not negligible because if someone is doing `prisma-import && prisma migrate dev` (which I do in my projects), they might create a migration that wipes half of the tables in their database. Again, I don't think there's an easy and unopinionated way of tackling this, so a log message saying that something like `Merged <schema1>, <schema2> and <schema3> schemas from <base>` might suffice.

nawilson commented on packages/prisma-import/src/index.ts:
```diff
@@ -40,43 +43,76 @@ ${chalk.bold('Flags')}
+    if (relativeBasePath) {
+      const absoluteBasePath = path.isAbsolute(relativeBasePath) ? relativeBasePath : path.resolve(relativeBasePath)
+
+      schemaPaths = await pathsFromBase(absoluteBasePath)
```
Added log.
Also added same logging to `--schemas` for consistency.
Also added `--quiet` flag in case developer wants logging off

nawilson commented:
Ok I understand now. Yes I agree that it will be up to the developer to ensure that the resulting `schema.prisma` is not missing any thing that should be included. I think the cli can facilitate this by providing what models, if any, are being removed when compared against a pre-existing `schema.prisma`. If models are removed, it can halt execution unless `--force` was supplied.

I can implement. However I would like to refactor to make better use of the existing code and ability to add jest tests.
Let me know if you're ok with that.

nawilson commented:
I've added additional logging to show what has been imported and a flag to turn it on: `--verbose`
I've also refactored the parsing logic to be separate and added some tests for it.

nawilson commented:
@ajmnz any update on how you want to handle this feature? Or any other revisions you would like made before it can be merged?

ajmnz commented:
@nawilson, sorry for the delay, I'm a bit busy these days.

Will review shortly. Thanks
